onnxruntime/core/language_interop_ops/torch/custom_function_register.cc

```diff
@@ -0,0 +1,261 @@
+  UnregisterEntry(&forward_runner);
+}
+
+void OrtTorchFunctionPool::UnregisterBackwardRunner() {
+  std::lock_guard<std::mutex> lock(func_context_pool_mutex_);
+  UnregisterEntry(&backward_runner);
+}
+
+PyObject* OrtTorchFunctionPool::GetForwardRunner() {
+  std::lock_guard<std::mutex> lock(func_context_pool_mutex_);
+  ORT_ENFORCE(forward_runner, "Forward runner cannot be NULL. Do you forget register it by calling RegisterForwardRunner(...)?");
+  return forward_runner;
+}
+
+PyObject* OrtTorchFunctionPool::GetBackwardRunner() {
+  std::lock_guard<std::mutex> lock(func_context_pool_mutex_);
+  ORT_ENFORCE(backward_runner, "backward runner cannot be NULL. Do you forget register it by calling RegisterBackwardRunner(...)?");
+  return backward_runner;
+}
+
+PyObject* OrtTorchFunctionPool::GetForwardCore(const std::string& key) {
+  ORT_ENFORCE(!key.empty(), "Cannot be empty string.");
+  std::lock_guard<std::mutex> lock(func_context_pool_mutex_);
+  auto iter = forward_core_pool.find(key);
+  ORT_ENFORCE(iter != forward_core_pool.end(), "No forward registered for ", key);
+  return iter->second;
+}
+
+PyObject* OrtTorchFunctionPool::GetBackwardCore(const std::string& key) {
+  ORT_ENFORCE(!key.empty(), "Cannot be empty string.");
+  std::lock_guard<std::mutex> lock(func_context_pool_mutex_);
+  auto iter = backward_core_pool.find(key);
+  ORT_ENFORCE(iter != backward_core_pool.end(), "No backward registered for ", key);
+  return iter->second;
+}
+
+int64_t OrtTorchFunctionPool::RegisterContext(PyObject* auto_grad_context) {
+  static int64_t index_ = 0x1000000;
+  std::lock_guard<std::mutex> lock(func_context_pool_mutex_);
+  index_++;
+#ifndef NDEBUG
+  RefCountTracker::GetInstance().TrackPyObject(RefCountTracker::ObjCategory::AutoGradContext,
+                                               auto_grad_context, "autograd_context_register");
+#endif
+  func_context_pool.insert({index_, auto_grad_context});
+  // We don't need increase the context refcnt because PyTorch already did it during .apply().
+  return index_;
+}
+
+void OrtTorchFunctionPool::UnRegisterContext(int64_t context_index) {
+  std::lock_guard<std::mutex> lock(func_context_pool_mutex_);
+  auto it = func_context_pool.find(context_index);
+
+  // We just need remove the context key value pair, the context itself
+  // will be removed, when forward outputs are destoyed.
+  ORT_ENFORCE(it != func_context_pool.end(),
+              "Cannot unregister unexisting key: ", context_index);
+  func_context_pool.erase(it);
+}
+
+PyObject* OrtTorchFunctionPool::GetContext(int64_t context_index) {
+  std::lock_guard<std::mutex> lock(func_context_pool_mutex_);
+  auto iter = func_context_pool.find(context_index);
+  ORT_ENFORCE(iter != func_context_pool.end(), "No context registered for ", context_index);
+  return iter->second;
+}
+
+template <typename TKey>
+void UnregisterAllEntries(
+    std::unordered_map<TKey, PyObject*>& pool) {
+  auto it = pool.begin();
+  while (it != pool.end()) {
```

Comment:
I don't think so. Since we are deleting elements in `pool`, the operatation `++it` may not be the actual `next` element after calling `pool.erase(...)`.